File: scripts/process-stickers.py
# Clean the raw sticker PNGs for the hero using ML background removal.
#
# ChatGPT bakes its "transparency" checkerboard into the PNG as real grey
# pixels, so a colour flood-fill can't tell the background from the light
# fills inside the artwork and tunnels into it. rembg (U^2-Net salient-object
# segmentation) understands the *subject* instead of the colour, so it lifts
# the whole sticker cleanly, holographic fills and interior features intact.
#
#   python3 -m venv .venv-rembg
#   source .venv-rembg/bin/activate
#   pip install "rembg" onnxruntime pillow numpy
#   python scripts/process-stickers.py
import io
import os
from pathlib import Path

import numpy as np
from PIL import Image, ImageFilter
from rembg import new_session, remove
from scipy import ndimage

RAW_DIR = Path("public/stickers/_raw")
OUT_DIR = Path("public/stickers")
FINAL = 384          # exported size
ALPHA_CUTOFF = 12    # alpha below this is treated as fully transparent for trim

# isnet-general-use gives cleaner edges than the default u2net for line-art.
session = new_session("isnet-general-use")


def solid_mask(raw: bytes) -> Image.Image:
    """Return a clean binary alpha for the whole sticker.

    rembg reliably finds the coloured artwork + dark outline but treats
    enclosed light regions (the pen-nib holographic fill, the mushroom cap,
    smiley eyes) as background, punching holes. We therefore take the raw
    mask, fill every interior hole so those regions become solid, keep only
    the largest blob (drops stray specks), and dilate a touch to recover the
    near-white die-cut border that colour-matches the background.

    Real die-cut holes (e.g. the pen-nib centre) are then re-opened: a filled
    hole is punched back out only when it is small AND its original colour
    matches the background — so genuine holes go transparent while black
    features (eyes) and the large holographic fill stay solid.
    """
    m = remove(raw, session=session, only_mask=True, post_process_mask=False)
    mask = np.array(Image.open(io.BytesIO(m)).convert("L"))
    rgb = np.array(Image.open(io.BytesIO(raw)).convert("RGB").resize(
        (mask.shape[1], mask.shape[0]), Image.LANCZOS
    ))
    binary = mask > 30

    filled = ndimage.binary_fill_holes(binary)

    # Identify genuine die-cut holes to punch back out. A true hole (the
    # pen-nib centre) is small, background-coloured, AND ringed by the dark
    # die-cut outline — unlike white specular glints, which sit small and
    # background-coloured but are surrounded by coloured/light fill.
    reopen = np.zeros_like(binary)
    holes = filled & ~binary
    if holes.any():
        k = 8
        corners = np.concatenate([
            rgb[:k, :k].reshape(-1, 3), rgb[:k, -k:].reshape(-1, 3),
            rgb[-k:, :k].reshape(-1, 3), rgb[-k:, -k:].reshape(-1, 3),
        ])
        bg = np.median(corners, axis=0)
        lum = rgb @ np.array([0.299, 0.587, 0.114])
        labels_h, n_h = ndimage.label(holes)
        total = float(binary.sum())
        for h in range(1, n_h + 1):
            comp = labels_h == h
            if comp.sum() > 0.08 * total:  # large hole (holo fill) → keep solid
                continue
            med = np.median(rgb[comp], axis=0)
            if np.linalg.norm(med - bg) >= 45:  # not background-coloured
                continue
            ring = ndimage.binary_dilation(comp, iterations=6) & ~comp
            if ring.any() and np.median(lum[ring]) < 70:  # ringed by dark outline
                reopen |= comp

    binary = filled
    # keep the largest connected component only
    labels, n = ndimage.label(binary)
    if n > 1:
        sizes = ndimage.sum(binary, labels, range(1, n + 1))
        binary = labels == (int(np.argmax(sizes)) + 1)

    # bridge grunge gaps, then grow slightly to catch the white border
    binary = ndimage.binary_closing(binary, iterations=4)
    binary = ndimage.binary_fill_holes(binary)
    binary = ndimage.binary_dilation(binary, iterations=3)

    # punch genuine die-cut holes back out
    binary = binary & ~reopen

    alpha = np.where(binary, 255, 0).astype(np.uint8)
    img = Image.fromarray(alpha, mode="L")
    # soften the hard edge by 1px for clean anti-aliasing
    return img.filter(ImageFilter.GaussianBlur(1.0))


def trim_to_content(img: Image.Image) -> Image.Image:
    """Crop to the alpha bounding box so the sticker fills the frame."""
    alpha = np.array(img.getchannel("A"))
    ys, xs = np.where(alpha > ALPHA_CUTOFF)
    if len(xs) == 0:
        return img
    x0, x1 = xs.min(), xs.max() + 1
    y0, y1 = ys.min(), ys.max() + 1
    return img.crop((x0, y0, x1, y1))


def fit_square(img: Image.Image, size: int) -> Image.Image:
    """Scale to fit inside a transparent square canvas, centred."""
    img.thumbnail((size, size), Image.LANCZOS)
    canvas = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    canvas.paste(img, ((size - img.width) // 2, (size - img.height) // 2), img)
    return canvas


sources = sorted(p for p in RAW_DIR.iterdir() if p.suffix.lower() == ".png")

for idx, src in enumerate(sources, start=1):
    with open(src, "rb") as fh:
        raw = fh.read()

    rgb = Image.open(io.BytesIO(raw)).convert("RGB")
    alpha = solid_mask(raw).resize(rgb.size, Image.LANCZOS)
    img = rgb.convert("RGBA")
    img.putalpha(alpha)

    img = trim_to_content(img)
    img = fit_square(img, FINAL)

    name = f"sticker-{idx:02d}.png"
    img.save(OUT_DIR / name, optimize=True)
    print("wrote", name)

print("done ->", OUT_DIR)

File: scripts/process-stickers.mjs
// One-off: clean the raw sticker PNGs for the hero.
//  1. Flood the background to transparent (already-transparent + an opaque
//     backdrop colour sampled from the corners).
//  2. From that background frontier, eat only a SHALLOW ring of near-black
//     die-cut outline (depth-limited + saturation-guarded) so it can't tunnel
//     into holographic / coloured fills or interior features (eyes, smile).
// Outputs sticker-01.png … in public/stickers.
//
//   node scripts/process-stickers.mjs
import sharp from "sharp";
import { readdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";

const DIR = "public/stickers";
const WORK = 640; // processing resolution
const FINAL = 384; // exported size (renders ~50px, so plenty sharp)
const BG_TOL = 42; // opaque backdrop colour tolerance
const DARK_LUM = 60; // outline is near-black
const DARK_SAT = 0.32; // …and near-grey (protects saturated fills)
const MAX_DARK_DEPTH = 26; // how far the black ring may be eaten (px @ WORK)

const sources = readdirSync(DIR)
  .filter((f) => /^ChatGPT.*\.png$/i.test(f))
  .sort();

for (let idx = 0; idx < sources.length; idx += 1) {
  const { data, info } = await sharp(join(DIR, sources[idx]))
    .ensureAlpha()
    .resize({ width: WORK, height: WORK, fit: "inside" })
    .raw()
    .toBuffer({ resolveWithObject: true });

  const { width: W, height: H, channels: C } = info;

  // Sample an opaque backdrop colour from the border pixels (if any).
  let sr = 0;
  let sg = 0;
  let sb = 0;
  let sn = 0;
  const sampleBorder = (x, y) => {
    const i = (y * W + x) * C;
    if (data[i + 3] > 200) {
      sr += data[i];
      sg += data[i + 1];
      sb += data[i + 2];
      sn += 1;
    }
  };
  for (let x = 0; x < W; x += 1) {
    sampleBorder(x, 0);
    sampleBorder(x, H - 1);
  }
  for (let y = 0; y < H; y += 1) {
    sampleBorder(0, y);
    sampleBorder(W - 1, y);
  }
  const bgOpaque = sn > W; // a decent chunk of the border is opaque
  const br = bgOpaque ? sr / sn : 0;
  const bg = bgOpaque ? sg / sn : 0;
  const bb = bgOpaque ? sb / sn : 0;

  const lumAt = (i) => 0.299 * data[i] + 0.587 * data[i + 1] + 0.114 * data[i + 2];
  const satAt = (i) => {
    const max = Math.max(data[i], data[i + 1], data[i + 2]);
    const min = Math.min(data[i], data[i + 1], data[i + 2]);
    return max === 0 ? 0 : (max - min) / max;
  };
  const isBg = (i) => {
    if (data[i + 3] < 25) return true;
    return (
      bgOpaque &&
      Math.abs(data[i] - br) < BG_TOL &&
      Math.abs(data[i + 1] - bg) < BG_TOL &&
      Math.abs(data[i + 2] - bb) < BG_TOL
    );
  };
  const isOutline = (i) => lumAt(i) < DARK_LUM && satAt(i) < DARK_SAT;

  // BFS carrying a "dark depth"; background resets it to 0, dark increments it.
  const visited = new Uint8Array(W * H);
  const queue = [];
  const depths = [];
  const consider = (x, y, depth) => {
    if (x < 0 || y < 0 || x >= W || y >= H) return;
    const p = y * W + x;
    if (visited[p]) return;
    const i = p * C;
    if (isBg(i)) {
      visited[p] = 1;
      queue.push(p);
      depths.push(0);
    } else if (isOutline(i) && depth < MAX_DARK_DEPTH) {
      visited[p] = 1;
      queue.push(p);
      depths.push(depth + 1);
    }
  };
  for (let x = 0; x < W; x += 1) {
    consider(x, 0, 0);
    consider(x, H - 1, 0);
  }
  for (let y = 0; y < H; y += 1) {
    consider(0, y, 0);
    consider(W - 1, y, 0);
  }
  let head = 0;
  while (head < queue.length) {
    const p = queue[head];
    const depth = depths[head];
    head += 1;
    data[p * C + 3] = 0;
    const x = p % W;
    const y = (p / W) | 0;
    consider(x + 1, y, depth);
    consider(x - 1, y, depth);
    consider(x, y + 1, depth);
    consider(x, y - 1, depth);
  }

  const out = await sharp(data, { raw: { width: W, height: H, channels: C } })
    .png()
    .trim({ threshold: 12 })
    .resize({ width: FINAL, height: FINAL, fit: "inside" })
    .toBuffer();

  const name = `sticker-${String(idx + 1).padStart(2, "0")}.png`;
  writeFileSync(join(DIR, name), out);
  console.log("wrote", name, "bgOpaque:", bgOpaque);
}
